ops/pebble.py

```diff
@@ -660,6 +660,22 @@ def to_dict(self) -> typing.Dict[str, typing.Any]:
         ]
         return {name: value for name, value in fields if value}
 
+    def _merge(self, other: 'Service'):
+        """Merges this service object with another service definition.
+
+        For attributes present in both objects, the passed in service
+        attributes take precedence.
+        """
+        for name, value in other.__dict__.items():
```

Comment:
This introduces some potential bugs.

`to_dict()` returns a subset of fields in the object, omitting fields like "name", which aren't part of the values that we want to override.

`__dict__` just returns the built in Python dictionary, discarding the Framework's attempt to exclude values that shouldn't be included.

Note that to_dict _always_ excludes empty values, which means that you don't need the first check. 

Ben is right about the dashes vs underscores. You need to transform those values.

Reply:
I think perhaps just skipping the `name` attrib would be good enough?  Then we don't have to do the underscore/dash stuff.

Reply:
@rwcarlsen that risks introducing more bugs in the future, as this object changes. For example, if somebody adds a helper method to the class, that will end up in the __dict__, and we might end up doing very weird things to it.

I would be on board with overriding the class's built in set and get with methods that replace the underscores, though that is getting into "too cute and too clever" territory :-)

Reply:
Fair enough.  Let's not override the built-in set/get though since it does indeed seem to be getting to clever.

Reply:
I did check that, and `__dict__` only returns the actual attributes `__init__ has set, not helper methods and the like. For example, it doesn't include `to_dict`. Note that the fact that it includes `name` is okay, because when merging services, the `name` attributes should always be the same (but I'm happy to exclude it explicitly if you think that's better).

```
>>> sorted(s.__dict__)
['after', 'backoff_delay', 'backoff_factor', 'backoff_limit', 'before', 'command', 'description', 'environment', 'group', 'group_id', 'name', 'on_check_failure', 'on_failure', 'on_success', 'override', 'requires', 'startup', 'summary', 'user', 'user_id']
```

I don't see why we should use `to_dict` and "transform" those keys back, that seems like an unnecessary transformation and round trip -- why not just use the attribute names directly in the first place?

That said, I'd be happy to have an explicit list of attributes in `_merge` if you think that's cleaner than using `__dict__`.

Reply:
I think @pengale's concern was that future additions to the Service class may be "special" in that just overwriting from one object to another could become problematic.  In theory this is a concern, but in practice I'm not sure how much it really matters - particularly since this is for testing and isn't strictly a matter of normal public APIs.

The entire point of the dynamic approach we've taken here is to have things auto-handle future changes.  An explicit list basically undoes those gains.  I think I'd lean toward using `__dict__`.

Reply:
so.. as you was proposed in the request for changes, i changed the method to use to_dict.. and then following the discussion, i changed it back to using __dict__..   but feel free to discuss this topic and let me know which you prefer.